pkgs/misc/uboot/hardkernel-firmware.nix

```diff
@@ -0,0 +1,142 @@
+{ stdenv
+, lib
+, fetchpatch
+, fetchFromGitHub
+, buildPackages
+, pkgsCross
+}:
+
+let
+  buildHardkernelFirmware = {
+    version ? null
+    , src ? null
+    , name ? ""
+    , filesToInstall
+    , installDir ? "$out"
+    , defconfig
+    , extraMeta ? {}
+    , ... } @ args: stdenv.mkDerivation ({
+      pname = "uboot-hardkernel-firmware-${name}";
+
+      nativeBuildInputs = [
+        buildPackages.git
+        buildPackages.hostname
+        pkgsCross.arm-embedded.stdenv.cc
+      ];
+
+      depsBuildBuild = [
+        buildPackages.gcc49
+      ] ++ lib.optional (stdenv.buildPlatform != stdenv.hostPlatform) buildPackages.stdenv.cc
+        ++ lib.optional (!stdenv.isAarch64) pkgsCross.aarch64-multiplatform.buildPackages.gcc49;
+
+      postPatch = ''
+        substituteInPlace Makefile --replace "/bin/pwd" "pwd"
+      '';
+
+      makeFlags = [
+        "CROSS_COMPILE=${stdenv.cc.targetPrefix}"
+        "CROSS_COMPILE_32=${pkgsCross.arm-embedded.stdenv.cc.targetPrefix}"
+        "${defconfig}" "bl301.bin"
+      ]
+      ++ lib.optional (!stdenv.isAarch64) "CROSS_COMPILE=${pkgsCross.aarch64-multiplatform.stdenv.cc.targetPrefix}";
+
+      installPhase = ''
+        mkdir -p ${installDir}
+        cp ${lib.concatStringsSep " " filesToInstall} ${installDir}
+      '';
+
+      meta = with lib; {
+        homepage = "https://www.hardkernel.com/";
+        description = "Das U-Boot from Hardkernel with Odroid embedded devices firmware and support";
+        license = licenses.unfreeRedistributableFirmware;
+        maintainers = with maintainers; [ aarapov ];
+      } // extraMeta;
+    } // removeAttrs args [ "extraMeta" ]);
+in {
+  inherit buildHardkernelFirmware;
+
+  firmwareOdroidC2 = buildHardkernelFirmware {
+    defconfig = "odroidc2_config";
+    name = "firmware-odroid-c2";
+    version = "2015.01";
+    src = fetchFromGitHub {
+      owner = "hardkernel";
+      repo = "u-boot";
+      rev = "fac4d2da0a1b61dfdeaca0034a45151ff5983fb8";
+      sha256 = "09s0y69ilrwnvqi1g11axsnhylq8kfljwqxdfjifa227mi0kzq37";
+    };
+
+    # https://wiki.odroid.com/odroid-c2/software/building_u-boot
+    prePatch = ''
+      patch -p1 < ${fetchpatch {
```

Comment:
I guess this was done to match the instructions given in that link, which tell you to use `git merge`, but it might be clearer and easier to maintain (e.g. you wouldn't need to add more patches if hardkernel ever decides to update the firmware) if you were to fetch the `u-boot_firmware` repo and just copy the firmware directories into place in the u-boot tree.

Reply:
This is almost the way it was before the last commit. I did think about just making a copy over, though I decided to keep a little more control here. I don't anticipate there will be any other change as the last one is dated to Feb 2017. And if there will be I do want to know what it is.
That said it is not an issue to maintain it, and in case we need to add a patch - we have a control in a way we don't get anything creepy and this new change won't break anything unexpectedly.

Reply:
ok, fine with me